osquery/tables/system/windows/windows_eventlog.cpp

```diff
@@ -0,0 +1,204 @@
+
+#include <Windows.h>
+#include <winevt.h>
+
+#include <boost/algorithm/string.hpp>
+#include <boost/foreach.hpp>
+#include <boost/property_tree/json_parser.hpp>
+#include <boost/property_tree/ptree.hpp>
+#include <boost/property_tree/xml_parser.hpp>
+
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+#include <osquery/core/windows/wmi.h>
+#include <osquery/events/windows/windowseventlogparser.h>
+#include <osquery/sql/dynamic_table_row.h>
+#include <osquery/utils/conversions/join.h>
+#include <osquery/utils/conversions/split.h>
+#include <osquery/utils/conversions/windows/strings.h>
+
+namespace pt = boost::property_tree;
+
+namespace osquery {
+namespace tables {
+
+const std::string kEventLogXmlPrefix = "<QueryList><Query Id=\"0\">";
+const std::string kEventLogXmlSuffix = "</Query></QueryList>";
+
+const int kNumEventsBlock = 1024;
+
+void parseWelXml(QueryContext& context,
+                 std::wstring& xml_event,
+                 RowYield& yield) {
+  auto row = make_table_row();
+  pt::ptree propTree;
+  WELEvent windows_event;
+  auto status = parseWindowsEventLogXML(propTree, xml_event);
+  status = parseWindowsEventLogPTree(windows_event, propTree);
+
+  row["time"] = INTEGER(windows_event.osquery_time);
+  row["datetime"] = SQL_TEXT(windows_event.datetime);
+  row["channel"] = SQL_TEXT(windows_event.source);
+  row["provider_name"] = SQL_TEXT(windows_event.provider_name);
+  row["provider_guid"] = SQL_TEXT(windows_event.provider_guid);
+  row["eventid"] = INTEGER(windows_event.event_id);
+  row["task"] = INTEGER(windows_event.task_id);
+  row["level"] = INTEGER(windows_event.level);
+  row["pid"] = INTEGER(windows_event.pid);
+  row["tid"] = INTEGER(windows_event.tid);
+
+  row["keywords"] = SQL_TEXT(windows_event.keywords);
+  row["data"] = SQL_TEXT(windows_event.data);
+
+  if (context.hasConstraint("time_range", EQUALS)) {
+    auto time_range = context.constraints["time_range"].getAll(EQUALS);
+    row["time_range"] = SQL_TEXT(*time_range.begin());
+  } else {
+    row["time_range"] = SQL_TEXT("");
+  }
+
+  if (context.hasConstraint("timestamp", EQUALS)) {
+    auto timestamp = context.constraints["timestamp"].getAll(EQUALS);
+    row["timestamp"] = SQL_TEXT(*timestamp.begin());
+  }
+
+  yield(std::move(row));
+}
+
+void parseQueryResults(QueryContext& context,
+                       EVT_HANDLE queryResults,
+                       RowYield& yield) {
+  std::vector<EVT_HANDLE> events(kNumEventsBlock);
+  unsigned long numEvents = 0;
+
+  // Retrieve the events one block at a time
+  auto ret = EvtNext(
+      queryResults, kNumEventsBlock, events.data(), INFINITE, 0, &numEvents);
+
+  while (ret != FALSE) {
+    for (unsigned long i = 0; i < numEvents; i++) {
+      std::vector<wchar_t> renderedContent;
+      unsigned long renderedBuffSize = 0;
+      unsigned long renderedBuffUsed = 0;
+      unsigned long propCount = 0;
+      EvtRender(nullptr,
+                events[i],
+                EvtRenderEventXml,
```

Comment:
Hesitating about EvtRenderEventXml as flag vs EvtRenderEventValues here, since EvtRenderEventXml produces an XML string that is now having to be reparsed. EvtRenderEventXml is used more commonly when your final result is XML and you'll perform transform on it (say with XSLT). Here, EvtRenderEventValues will be faster and cleaner since you can avoid having to write parser code.

Reply:
Thanks @farfella for the pointers. I looked into `EvtRenderEventValues` and it will save the step of parsing the XML. The reason I choose to go with `EvtRenderEventXml` because the existing event-based WEL table uses the same approach and also getting the user-data or event-data is convenient from the XML format. 

I will take a note of your suggestions and have a future improvement with using `EvtRenderEventValues` flag.